File: src/layer1_wasm/php-12167/repro.php
<?php
// Vivarium Layer 1 reproduction — php/php-src#12167, native variant.
//
// Mirrors the script that runs in `repro.ts` (under php-wasm) so a
// contributor can re-verify the bug against a real PHP interpreter:
//
//   mise install               # one-time, picks up .mise.toml
//   mise exec php -- php src/layer1_wasm/php-12167/repro.php
//
// The script prints `pass` if the bug REPRODUCES (xpath returns the
// processing-instruction node but casting it to string yields ""),
// `fail` otherwise — same verdict semantics as the in-browser page.
// Exit code: 0 on `pass`, 1 on `fail` so CI can shell-script around
// it without parsing stdout.

$xml = '<?xml version="1.0"?><foo><bar><?stylesheet hello ?></bar></foo>';
$sxe = simplexml_load_string($xml);
$pis = $sxe->xpath("//processing-instruction()");
$pi_text = isset($pis[0]) ? (string) $pis[0] : null;

$result = [
    "php_version"   => PHP_VERSION,
    "xpath_count"   => count($pis ?: []),
    "pi_text"       => $pi_text,
    "pi_text_empty" => $pi_text === "",
];

$reproduced = $result["xpath_count"] === 1 && $result["pi_text_empty"];

echo json_encode($result, JSON_UNESCAPED_SLASHES | JSON_PRETTY_PRINT) . "\n";

if ($reproduced) {
    fwrite(STDOUT, "verdict=pass — bug reproduces on this interpreter\n");
    exit(0);
} elseif ($result["xpath_count"] === 1 && !$result["pi_text_empty"]) {
    fwrite(STDOUT, "verdict=fail — SimpleXML now returns the PI content (likely fixed upstream)\n");
    exit(1);
} else {
    fwrite(
        STDOUT,
        "verdict=fail — unexpected outcome (xpath_count={$result['xpath_count']}, pi_text=" . json_encode($result["pi_text"]) . ")\n",
    );
    exit(1);
}

File: src/layer1_wasm/ruby-21709/repro.rb
# Vivarium Layer 1 reproduction — ruby/ruby#21709, native variant.
#
# Mirrors the script that runs in `repro.ts` (under Ruby.wasm) so a
# contributor can re-verify the bug against a real Ruby interpreter:
#
#   mise install                     # one-time, picks up mise.toml
#   mise exec ruby -- ruby src/layer1_wasm/ruby-21709/repro.rb
#
# Interpolates the same two fragments — one UTF-8, one US-ASCII — into
# a Regexp and into a String, and prints what each one did. Exits 0 on
# `reproduced` (Regexp interpolation rejects the combine ∧ String
# interpolation accepts it), 1 on `unreproduced`, so CI can shell-script
# around it without parsing stdout.

prefix = '\p{In_Arabic}'
suffix = '\p{In_Arabic}'.encode('US-ASCII')

begin
  re = /#{prefix}#{suffix}/
  regexp_raised = nil
rescue => e
  regexp_raised = e.class.name
end

begin
  str = "#{prefix}#{suffix}"
  string_encoding = str.encoding.name
  string_raised = nil
rescue => e
  string_encoding = nil
  string_raised = e.class.name
end

$result = {
  ruby_version: RUBY_VERSION,
  regexp_built: regexp_raised.nil?,
  regexp_raised: regexp_raised,
  string_built: string_raised.nil?,
  string_encoding: string_encoding,
  string_raised: string_raised,
}

regexp_note =
  regexp_raised ? "raised #{regexp_raised}   <-- rejected" : 'built'
string_note =
  string_raised ? "raised #{string_raised}" : "built, encoding #{string_encoding}"

puts 'Interpolating the same two fragments, one UTF-8 and one US-ASCII:'
puts
puts '  Regexp   /#{prefix}#{suffix}/    ' + regexp_note
puts '  String   "#{prefix}#{suffix}"    ' + string_note
puts
if regexp_raised && string_raised.nil?
  puts 'The two forms disagree: Regexp interpolation rejects the mixed'
  puts 'encodings that String interpolation silently upgrades.'
else
  puts 'The two forms agree on how to combine fragments of different encodings.'
end
puts "Ruby #{RUBY_VERSION}"

if $result[:regexp_built] == false && $result[:string_built]
  warn 'verdict=reproduced — bug reproduces on this interpreter'
  exit 0
elsif $result[:regexp_built] && $result[:string_built]
  warn 'verdict=unreproduced — Regexp and String interpolation now agree (likely fixed upstream)'
  exit 1
else
  warn "verdict=unreproduced — unexpected outcome (regexp_built=#{$result[:regexp_built]}, string_built=#{$result[:string_built]})"
  exit 1
end
